src/shogun/kernel/string/WeightedDegreeStringKernel.cpp

```diff
@@ -137,6 +132,9 @@ void CWeightedDegreeStringKernel::create_empty_tries()
 
 bool CWeightedDegreeStringKernel::init(CFeatures* l, CFeatures* r)
 {
+	if (type != E_EXTERNAL)
+		set_wd_weights_by_type(type);
```

Comment:
to initialize weights based on type, since "put" changes the type quietly

Reply:
I think rather than checking which overloading of `init` is called, you can check whether features have been initialized via `put`. This will allow us to initialize kernels in this style, which has been already used in a few examples
```
k = kernel('WeightedDegreeStringKernel', lhs=..., rhs=...)
k.init()```

Reply:
the Kernel interface has only one init, which accepts lhs and rhs, do you mean Distance?
`‘void shogun::CKernel::init()’ is private within this context`

Reply:
Wuwei we marked lhs ans rhs as read only to force users to call kernel.init recently 
Avoids all sorts of problems
So this change is good

Reply:
Didn't know that, thanks for the clarification